crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,453 @@
+            multiline: is_multiline,
+        })
+    }
+
+    fn construct_sorted_all(
+        &self,
+        locator: &Locator,
+        parent: &ast::Stmt,
+        stylist: &Stylist,
+    ) -> SortedDunderAll {
+        let mut sorted_items = self.items.clone();
+        sorted_items.sort();
+        if sorted_items == self.items {
+            return SortedDunderAll {
+                was_already_sorted: true,
+                new_dunder_all: None,
+            };
+        }
+        // As well as the "items" in the `__all__` definition,
+        // there is also a "prelude" and a "postlude":
+        //  - Prelude == the region of source code from the opening parenthesis
+        //    (if there was one), up to the start of the first element in `__all__`.
+        //  - Postlude == the region of source code from the end of the last
+        //    element in `__all__` up to and including the closing parenthesis
+        //    (if there was one).
+        let prelude_end = {
+            if let Some(first_item) = self.items.first() {
+                let first_item_line_offset = locator.line_start(first_item.start());
+                if first_item_line_offset == locator.line_start(self.start()) {
+                    first_item.start()
+                } else {
+                    first_item_line_offset
+                }
+            } else {
+                self.start() + TextSize::new(1)
+            }
+        };
+        let postlude_start = {
+            if let Some(last_item) = self.items.last() {
+                let last_item_line_offset = locator.line_end(last_item.end());
+                if last_item_line_offset == locator.line_end(self.end()) {
+                    last_item.end()
+                } else {
+                    last_item_line_offset
+                }
+            } else {
+                self.end() - TextSize::new(1)
+            }
+        };
+        let mut prelude = locator
+            .slice(TextRange::new(self.start(), prelude_end))
+            .to_string();
+        let postlude = locator.slice(TextRange::new(postlude_start, self.end()));
+
+        let joined_items = if self.multiline {
+            let indentation = stylist.indentation();
+            let newline = stylist.line_ending().as_str();
+            prelude = format!("{}{}", prelude.trim_end(), newline);
+            join_multiline_dunder_all_items(&sorted_items, locator, parent, indentation, newline)
+        } else {
+            Some(join_singleline_dunder_all_items(&sorted_items, locator))
+        };
+
+        let new_dunder_all = joined_items.map(|items| format!("{prelude}{items}{postlude}"));
+        SortedDunderAll {
+            was_already_sorted: false,
+            new_dunder_all,
+        }
+    }
+}
+
+impl Ranged for DunderAllValue<'_> {
+    fn range(&self) -> TextRange {
+        *self.range
+    }
+}
+
+struct SortedDunderAll {
+    was_already_sorted: bool,
+    new_dunder_all: Option<String>,
+}
```

Comment:
This could be an `enum` e.g. (not necessarily the best names)

```rust
enum SortedDunderAll {
    AlreadySorted,
    Sorted(Option<String>)
}
```

Reply:
Oh nice! Will make that change

Reply:
Alternatively, `construct_sorted_all` could return `Option<SortedDunderAll>`, where `None` indicates "already sorted". It's the same idea. `enum SortedDunderAll` is arguably more explicit.

Reply:
There's three states we need to represent here:
- `__all__` was already sorted; don't emit the violation
- `__all__` was not sorted, so we should emit the violation, but we don't know how to safely fix it here for whatever reason; something failed along the way
- `__all__` was not sorted, and here's the fix

For now I've gone with the enum, which I think is definitely better than what I originally had, but the naming still doesn't feel _ideal_. Open to suggestions :)

Reply:
Oops, good call!

Reply:
Added some more docs on this in my latest push, so I'll mark this as "resolved" for now, but still open to better naming suggestions!